Allow `_gce_` network when not using discovery gce

dadoonet opened the pull request:
For now we support `_gce_` only if discovery is set to `gce` and all information about GCE is provided (project_id and zone).
But in some cases, people would like to only bind to `_gce_` on a single node (without any elasticsearch cluster).

They could access the machine then from other machines running inside the same project.

This commit adds a new GceMetadataService which is started as soon as the plugin is started so GceNameResolver can use it to resolve `_gce`.

Closes #15724.

(cherry picked from commit 9871e41)

dadoonet commented:
@imotov As discussed in #15726 I cherry picked my code and applied it to master branch.
@bleskes Could you review this please?

bleskes commented:
I think we if we break GceComputeService into two functions (network + discovery) we should have those clearly named. It seems GceComputeService needs to become GceDiscoveryService in it's current implementation.

bleskes commented on plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceModule.java:
```diff
@@ -20,17 +20,37 @@
     @Override
     protected void configure() {
-        bind(GceComputeService.class).to(computeServiceImpl).asEagerSingleton();
+        if (isDiscoveryAlive(settings, logger)) {
```
why do we need this and the code in plugin.java?

```
         if (isDiscoveryAlive(settings, logger)) {
              services.add(GceModule.getComputeServiceImpl());
          }
         services.add(GceModule.getMetadataServiceImpl());
```

bleskes commented on plugins/discovery-gce/src/main/java/org/elasticsearch/plugin/discovery/gce/GceDiscoveryPlugin.java:
```diff
@@ -118,9 +117,9 @@ public static boolean isDiscoveryAlive(Settings settings, ESLogger logger) {
-        if (checkProperty(GceComputeService.Fields.PROJECT, settings.get(GceComputeService.Fields.PROJECT), logger) == false ||
-                checkProperty(GceComputeService.Fields.ZONE, settings.getAsArray(GceComputeService.Fields.ZONE), logger) == false) {
-            logger.debug("one or more gce discovery settings are missing. " +
+        if (checkProperty(settings.get(GceComputeService.Fields.PROJECT)) == false ||
```
if someone configures discovery.type to GCE and these properties are missing, we should throw a hard exception no? why do we only warn?

bleskes commented on plugins/discovery-gce/src/main/java/org/elasticsearch/plugin/discovery/gce/GceDiscoveryPlugin.java:
```diff
@@ -132,20 +131,12 @@ public static boolean isDiscoveryAlive(Settings settings, ESLogger logger) {
-            return false;
-        }
-        return true;
+    private static boolean checkProperty(String value) {
```
I don't think this method adds value any more?

bleskes replied:
oh, I see - it's a variant of the array. can we name this isNonEmpty?

dadoonet commented:
@bleskes Sorry it took me a loooong time to apply your changes...

> I think we if we break GceComputeService into two functions (network + discovery) we should have those clearly named. It seems GceComputeService needs to become GceDiscoveryService in it's current implementation.

I don't think we should change the names here. The two service classes we have here are responsible for:

* Calling GCE API (GceComputeService)
* Getting GCE instance metadata (GceMetadataService)

The discovery is done in `GceUnicastHostsProvider` which uses `GceComputeService` to get the list of the IP addresses running in GCE.

WDYT?

dadoonet commented:
@bleskes Before you can review it, I need to merge the master changes in it first. Some code changed. I'll ping you when ready to review.

dakrone commented:
@dadoonet is this still waiting on changes in master?

dadoonet commented:
@bleskes I finally ended up rebasing everything on master so it's ready for review and hopefully merged! :)

Thanks for your patience!

dadoonet commented:
@bleskes Do you have time to review this again?

bleskes commented on plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceComputeServiceImpl.java:
```diff
@@ -168,17 +127,17 @@ public GenericUrl run() {
         this.metaDataUrl =  gceHost + "/computeMetadata/v1/instance";
         this.gceRootUrl = GCE_ROOT_URL.get(settings);
-        tokenServerEncodedUrl = metaDataUrl + "/service-accounts/default/token";
+        this.tokenServerEncodedUrl = metaDataUrl + "/service-accounts/default/token";
```
why do we need all these urls if we have a dedicated service to fetch meta data? sounds duplicate. Also GceComputeService is a duplicate - the compute is already part of the Gce prefix. Since this is only in charge of resolving  instances of the same project, maybe call it GceProjectService or GceInstancesService?

dadoonet replied:
Makes sense. Done.

bleskes commented on plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceMetadataServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+    //      http://metadata/computeMetadata/v1/instance/service-accounts/default/token
+    // See https://developers.google.com/compute/docs/metadata#metadataserver
+    // all settings just used for testing - not registered by default
+    public static final Setting<String> GCE_HOST =
```
this is defined twice? there is also one on the ComputeService (which probably can go away)?

dadoonet replied:
++

bleskes commented on plugins/discovery-gce/src/test/java/org/elasticsearch/discovery/gce/GceMetadataServiceMock.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Mock for GCE Metadata Service
+ */
+public class GceMetadataServiceMock extends GceMetadataServiceImpl {
```
++ for this setup, but it makes the GceMetadataService interface redundant, right? we can just call GceMetadataServiceImpl  GceMetadataService and remove the interface, right ?

dadoonet replied:
Correct.

bleskes commented on plugins/discovery-gce/src/test/java/org/elasticsearch/discovery/gce/GceMockUtils.java:
```diff
@@ -0,0 +1,98 @@
+        // We extract from the url the mock file path we want to use
+        String mockFileName = Strings.replace(url, urlRoot, "");
+
+        URL resource = GceComputeServiceMock.class.getResource(mockFileName);
```
why do we refer to another class here?

dadoonet replied:
good catch. Bad copy and paste

bleskes commented on plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceMetadataServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+    @Inject
+    public GceMetadataServiceImpl(Settings settings, NetworkService networkService) {
+        super(settings);
+        networkService.addCustomNameResolver(new GceNameResolver(settings, this));
```
can we flip this around and not make GceMetadataService depend on NetworkService? rather, we can remove this from guice and instantiate it in the plugin and use it to create the GceNameResolver.

dadoonet replied:
I wonder how can I access to the NetworkService instance in the plugin class? Any idea?

bleskes replied:
I got confused expecting this to be the NetworkModule as we typically inject stuff on the module level. Looking at it, I think we should do it like this (it will require a separate change in core) - currently we create the NetworkService externally and pass it along to the NetworkModule. However, we never use that NetworkService anywhere else.

Instead, we should move the custom name resolver registration to the NetworkModule and pass the final list to the constructor of NetworkService , which can be created in NetworkModule#configure(). WDYT?

bleskes commented:
thx @dadoonet . I left some comments. please ping if something needs clarification

dadoonet commented on plugins/discovery-gce/src/main/java/org/elasticsearch/cloud/gce/GceMetadataServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Override
+    public String metadata(String metadataPath) throws IOException {
+        String urlMetadataNetwork = this.metaDataUrl + "/" + metadataPath;
```
Sure! That's better for sure.

dadoonet commented:
Thanks a lot @bleskes!
I pushed some other commits to address your remarks.
I need your help with this one I was not able to apply: https://github.com/elastic/elasticsearch/pull/15765#discussion_r65528384

dadoonet commented:
@bleskes I just added a new commit. (yeah I know: it took time) 

I was not able for now to add the network name resolver in the NetworkService ctor. 
The NetworkService is built in Node class with:

```java
            final NetworkService networkService = new NetworkService(settings);
```

So I'm unsure on how I could deal with that.

For now, I added a method which can add to the existing network service the network name resolver.

Let me know what you think.

dadoonet commented:
@rjernst I added this 5e57febe53c6961b65701d3f74e6af91559018d4 to address your comment.
I think I will come with a follow up PR to also remove:

```java
    public void onModule(DiscoveryModule discoveryModule) {
        discoveryModule.addDiscoveryType(EC2, ZenDiscovery.class);
        discoveryModule.addUnicastHostProvider(EC2, AwsEc2UnicastHostsProvider.class);
    }
```

And add new method in the DiscoveryPlugin interface instead.

rjernst commented:
@dadoonet Thanks for trying to move this to the new plugin model. But I don't think we should leave it in a state of adding a new use of onModule unless there is a blocker issue for the next release.

dadoonet commented:
@rjernst I'm not adding a new use of `onModule` here. It was there already and I don't touch it (yet).
I think this should come as a follow up for this PR.

Do you think we should block this PR until we changed the Discovery plugins to the new model? I gave a first look at what is needed to achieve this goal and it looks a bit complicated. 

That's why I'd prefer merging this PR first. I opened it in January and would love to have it merged.

WDYT?

dadoonet commented:
@rjernst Thanks for the advice. Makes a lot of sense. I pushed a new commit.

rjernst commented on core/src/main/java/org/elasticsearch/common/network/NetworkService.java:
```diff
@@ -273,4 +270,19 @@ public InetAddress resolvePublishHostAddresses(String publishHosts[]) throws IOE
+     * Register custom name resolver a DiscoveryPlugin might provide
+     * @param discoveryPlugins Discovery plugins
+     */
+    public static List<CustomNameResolver> registerCustomNameResolvers(Settings settings, List<DiscoveryPlugin> discoveryPlugins) {
```
This is an odd name (it doesnt actually register anything), and I don't see how it belongs in NetworkService (it has to do with how node construction creates the NetworkService). I think it should be in Node (and it can then be private).

dadoonet replied:
@rjernst Totally agreed. Was a left over.
About making it private, I can't because its also used by TransportClient.
May be I should move it in a NodeUtils class or something? Or better in `DiscoveryPlugin` interface? WDYT?

rjernst replied:
I don't think it should be in `DiscoveryPlugin`, since it is an implementation detail. Why is it used in the transport client? I did not think any of our cloud plugins were "transport client plugins"? (note that we currently do not have them marked for publication to maven, like we added for eg reindex)

dadoonet replied:
Argh! This is true.

This is also a left over. Sorry. My bad. Will fix.

rjernst commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -721,4 +723,19 @@ public static CircuitBreakerService createCircuitBreakerService(Settings setting
+     * Generate Custom Name Resolvers list based on a Discovery Plugins list
+     * @param discoveryPlugins Discovery plugins list
+     */
+    private List<NetworkService.CustomNameResolver> generateCustomNameResolvers(List<DiscoveryPlugin> discoveryPlugins) {
```
nit: generate implies creation, but really this is only pulling them all together. Perhaps just `get`?

rjernst commented on core/src/main/java/org/elasticsearch/plugins/DiscoveryPlugin.java:
```diff
@@ -0,0 +1,45 @@
+ */
+public interface DiscoveryPlugin {
+    /**
+     * Override to add additional {@link NetworkService.CustomNameResolver}s.
```
A description here on why someone would want a custom name resolve would be nice.

rjernst commented:
LGTM, I left 2 very minor additional comments.

dadoonet commented:
Thank you @rjernst! 

I also added a [latest commit](https://github.com/elastic/elasticsearch/pull/15765/commits/fb9bad23de1febf5ed84a9da286d27f7d0d1d7af ). I forgot to address one of @bleskes comment.

Do you mind giving a final review on this?

bleskes commented:
the two last two commits look good to me

dadoonet commented:
\o/ Fantastic guys! Thank you so much @rjernst and @bleskes for the help on this!
